app/src/main/java/protect/card_locker/LoyaltyCardViewActivity.java

```diff
@@ -520,6 +533,79 @@ private void showInfoDialog() {
         infoDialog.create().show();
     }
 
+    private void showBalanceUpdateDialog() {
+        AlertDialog.Builder builder = new MaterialAlertDialogBuilder(this);
+        builder.setTitle(R.string.updateBalanceTitle);
+        FrameLayout container = new FrameLayout(this);
+        FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(
+                ViewGroup.LayoutParams.MATCH_PARENT,
+                ViewGroup.LayoutParams.WRAP_CONTENT
+        );
+        params.leftMargin = 60;
+        params.rightMargin = 60;
+
+        LinearLayout layout = new LinearLayout(this);
+        layout.setOrientation(LinearLayout.VERTICAL);
+
+        TextView currentTextview = new TextView(this);
+        currentTextview.setText(getString(R.string.currentBalanceSentence, Utils.formatBalance(this, loyaltyCard.balance, loyaltyCard.balanceType)));
+        layout.addView(currentTextview);
+
+        TextView updateTextView = new TextView(this);
+        updateTextView.setText(getString(R.string.newBalanceSentence, Utils.formatBalance(this, loyaltyCard.balance, loyaltyCard.balanceType)));
+        layout.addView(updateTextView);
+
+        final EditText input = new EditText(this);
+        Context dialogContext = this;
+        input.setInputType(InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_FLAG_DECIMAL);
+        input.setHint(R.string.updateBalanceHint);
+        input.addTextChangedListener(new SimpleTextWatcher() {
+            @Override
+            public void onTextChanged(CharSequence s, int start, int before, int count) {
+                BigDecimal newBalance;
+                try {
+                    newBalance = calculateNewBalance(loyaltyCard.balance, loyaltyCard.balanceType, s.toString());
+                } catch (ParseException e) {
+                    updateTextView.setTag(null);
+                    updateTextView.setError(getString(R.string.parsingBalanceFailed));
+                    updateTextView.setText(getString(R.string.newBalanceSentence, Utils.formatBalance(dialogContext, loyaltyCard.balance, loyaltyCard.balanceType)));
+                    return;
+                }
+
+                // Save new balance into this element
+                updateTextView.setTag(newBalance);
+                updateTextView.setError(null);
+                updateTextView.setText(getString(R.string.newBalanceSentence, Utils.formatBalance(dialogContext, newBalance, loyaltyCard.balanceType)));
```

Comment:
The `setError` doesn't show the actual error message due to it being set on the `updateTextView` instead of on the `input` field the user is actually entering text in:

![image](https://user-images.githubusercontent.com/1885159/195416081-56ec9928-d653-4434-a45b-11a2d2e08f6e.png)

(Compared to correct on #1077)

However, from messing around a lot the only invalid state I could cause was the empty state, so maybe just only use `setTag` and completely remove `setError` after all because it doesn't look that great (I'd also prefer the `setTag` to be on the `input` itself instead of on the `updateTextView`, as the input contains the number the user entered, that way the `getTag` version would be the "sanitized" version).

```suggestion
                BigDecimal newBalance;
                try {
                    newBalance = calculateNewBalance(loyaltyCard.balance, loyaltyCard.balanceType, s.toString());
                } catch (ParseException e) {
                    input.setTag(null);
                    updateTextView.setText(getString(R.string.newBalanceSentence, Utils.formatBalance(dialogContext, loyaltyCard.balance, loyaltyCard.balanceType)));
                    return;
                }

                // Save new balance into this element
                input.setTag(newBalance);
                updateTextView.setText(getString(R.string.newBalanceSentence, Utils.formatBalance(dialogContext, newBalance, loyaltyCard.balanceType)));
```